### PR TITLE
Add Amazon Bedrock LLM endpoint integration

### LINKED ISSUE #775: Support for Amazon Bedrock
Hello,

I would like to request the addition of support for Amazon Bedrock to the Langchain library. As Amazon Bedrock is a new service, it would be beneficial for Langchain to include it as a supported platform.

2023-04-13 Amazon announced the new service [Amazon Bedrock](https://aws.amazon.com/bedrock/).
Blog: https://aws.amazon.com/blogs/machine-learning/announcing-new-tools-for-building-with-generative-ai-on-aws/

Issue of the python library: https://github.com/hwchase17/langchain/issues/2828

### PR DESCRIPTION
**Pull request by vitaly-ps**
A contribution from Prompt Security

* Kept as close to the Python implementation as possible
* Followed the guidelines from https://github.com/hwchase17/langchainjs/blob/main/CONTRIBUTING.md and https://github.com/hwchase17/langchainjs/blob/main/.github/contributing/INTEGRATIONS.md
* Covered with unit tests
* Added documentation

Fixes #775

### CODE REVIEW

File: docs/extras/modules/model_io/models/llms/integrations/bedrock.mdx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+## Installation and Setup
+
+Install the underlying library from AWS
+```bash npm2yarn
+npm install aws-sigv4-fetch
+```
+
+# LLM example usage
+
+```typescript
```
Comment (jacoblee93) [inline docs/extras/modules/model_io/models/llms/integrations/bedrock.mdx:14]: Just import this from `examples` so it's linted and formatted:

```
import CodeBlock from "@theme/CodeBlock";
import Example from "@examples/src/llms/bedrock.ts";

<CodeBlock language="typescript">{Example}</CodeBlock>
```

File: examples/src/models/llm/bedrock.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+// TODO:: add content here
```
Comment (jacoblee93) [inline examples/src/models/llm/bedrock.ts:1]: I think you have this above?

File: langchain/src/llms/bedrock.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        }: ${await response.text()}`
+      );
+    }
+    const responseJson = await response.json();
+    const text = LLMInputOutputAdapter.prepareOutput(provider, responseJson);
+    return text;
+  }
+
+  /** @ignore */
+  static async imports(): Promise<{ createSignedFetcher: any }> {
```
Comment (jacoblee93) [inline langchain/src/llms/bedrock.ts:144]: If you set up the entrypoints correctly (which it looks like you have), you can actually just put this at the top.

However, had a look at the package - it's not very popular and it doesn't look like it's doing much more than wrap the default AWS signature SDK. Could we just use that?

https://github.com/zirkelc/aws-sigv4-fetch/blob/main/src/create-signed-fetcher.ts#L14

**Comment by jacoblee93** [comment]
Looks good overall - a few comments!

File: docs/extras/modules/model_io/models/llms/integrations/bedrock.mdx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# Bedrock
+
+>[Amazon Bedrock](https://aws.amazon.com/bedrock/) is a fully managed service that makes FMs from leading AI startups and Amazon available via an API, so you can choose from a wide range of FMs to find the model that is best suited for your use case.
```
Comment (jacoblee93) [inline docs/extras/modules/model_io/models/llms/integrations/bedrock.mdx:3]: Would expand "FMs" to `Foundational Models (FMs)` the first time you use it

File: langchain/src/llms/bedrock.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
+import { getEnvironmentVariable } from "../util/env.js";
+import { LLM, BaseLLMParams } from "./base.js";
+
+type Dict = { [key: string]: any };
+
+class LLMInputOutputAdapter {
```
Comment (jacoblee93) [inline langchain/src/llms/bedrock.ts:6]: Nice! Can you rename this to `BedrockLLMInputOutputAdapter`?

File: langchain/src/llms/bedrock.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  /** The AWS region e.g. `us-west-2`.
+      Fallback to AWS_DEFAULT_REGION env variable or region specified in ~/.aws/config in case it is not provided here.
+  */
+  regionName?: string;
+
+  /** Temperature */
+  temperature?: number;
+
+  /** Max tokens */
+  maxTokens?: number;
```
Comment (jacoblee93) [inline langchain/src/llms/bedrock.ts:61]: We should also supply a way to pass in AWS credentials manually

File: langchain/src/llms/bedrock.ts (hunk 3 above)
Comment (jacoblee93) [reply langchain/src/llms/bedrock.ts:150]: Yeah, that's what I was thinking - they're MIT licensed but I worry about their long-term stability and I would rather depend directly on AWS' package.

Comment (jacoblee93) [reply langchain/src/llms/bedrock.ts:150]: That sounds good! I'm also fine with inlining and refactoring later.

Not sure what you mean by hard dependency though - if separated out correctly into an entrypoint one can import the rest of LangChain without importing this file and avoid relying on @aws-sdk/signature-v4. 

I can resolve any build issues at the end though, so don't worry about that too much.

**Comment by jacoblee93** [comment]
Hey @vitaly-ps looks great! Moved the hard deps to peer/optional deps but couldn't push more commits to your branch - I opened #2277 with those changes + some build and lint fixes. 

Can you take a look and test? If it looks good, we can merge tonight!

**Comment by mlconnor** [comment]
Curious if there needs to be a separate class for embeddings. I noticed that in Bedrock, embeddings use the same invoke endpoint with a specific model call, e.g. amazon.titan-e1t-medium.

**Comment by jacoblee93** [comment]
> > your branch
> 
> @jacoblee93 I've pulled your last changes from your feature-branch and included them here. Look at the last commits in this PR. Thanks.

Ok great! Will merge this one instead.

> Curious if there needs to be a separate class for embeddings. I noticed that in Bedrock, embeddings use the same invoke endpoint with a specific model call, e.g. amazon.titan-e1t-medium.

Yes, embeddings are a separate class.

**Comment by jacoblee93** [comment]
Thank you!

**Pull request merged**